jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +268,408 @@ public InvocationType getInvocationType()
+                else
+                {
+                    throw new IllegalStateException("Cannot overwrite exiting content " + c + " with " + content);
+                }
+            }
+        }
+
+        @Override
+        public boolean onTimeout(Throwable failure)
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c != DEMANDING)
+                    return false;
+                if (_content.compareAndSet(DEMANDING, new HttpInput.ErrorContent(failure)))
+                    return true;
+            }
+        }
+
+        @Override
+        public void eof()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == DEMANDING)
+                {
+                    if (_content.compareAndSet(DEMANDING, EOF))
+                        return;
+                }
+                else if (c == null)
+                {
+                    if (_content.compareAndSet(null, EOF))
+                        return;
+                }
+                else if (c.isEof())
+                {
+                    return;
+                }
+                else if (c.remaining() == 0)
+                {
+                    if (_content.compareAndSet(c, EOF))
+                        return;
+                }
+                else
+                {
+                    HttpInput.Content content = new HttpInput.Content(c.getByteBuffer())
+                    {
+                        @Override
+                        public boolean isEof()
+                        {
+                            return true;
+                        }
+
+                        @Override
+                        public void succeeded()
+                        {
+                            c.succeeded();
+                        }
+
+                        @Override
+                        public void failed(Throwable x)
+                        {
+                            c.failed(x);
+                        }
+
+                        @Override
+                        public InvocationType getInvocationType()
+                        {
+                            return c.getInvocationType();
+                        }
+                    };
+                    if (_content.compareAndSet(c, content))
+                        return;
+                }
+            }
+        }
+
+        @Override
+        public void failContent(Throwable failure)
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == null || c.isSpecial())
+                    return;
+                if (_content.compareAndSet(c, null))
+                {
+                    c.failed(failure);
+                    return;
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean needContent()
+    {
+        Tracer.in();
```

Comment:
Can we remove the Tracer from all code now

Reply:
Done.